### PR TITLE
BUG: pathlib.Path in io

### LINKED ISSUE #14932: pd.read_pickle, pd.DataFrame.to_pickle do not accept pathlib.Path objects
#### Code Sample

```python
In [1]: import pandas as pd
   ...: import pathlib as pl
   ...: path = pl.Path.home() / 'tmp.pickle' # I have prepared a pickle in this path
   ...: 

In [2]: df = pd.read_pickle(str(path)) # This works, no problem
   ...: 

In [3]: df2 = pd.read_pickle(path) # This breaks
   ...: 
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/home/ntaw/anaconda3/lib/python3.5/site-packages/pandas/io/pickle.py in try_read(path, encoding)
     50         try:
---> 51             with open(path, 'rb') as fh:
     52                 return pkl.load(fh)

TypeError: invalid file: PosixPath('/home/ntaw/tmp.pickle')

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
/home/ntaw/anaconda3/lib/python3.5/site-packages/pandas/io/pickle.py in try_read(path, encoding)
     55             try:
---> 56                 with open(path, 'rb') as fh:
     57                     return pc.load(fh, encoding=encoding, compat=False)

TypeError: invalid file: PosixPath('/home/ntaw/tmp.pickle')

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
/home/ntaw/anaconda3/lib/python3.5/site-packages/pandas/io/pickle.py in read_pickle(path)
     64     try:
---> 65         return try_read(path)
     66     except:

/home/ntaw/anaconda3/lib/python3.5/site-packages/pandas/io/pickle.py in try_read(path, encoding)
     60             except:
---> 61                 with open(path, 'rb') as fh:
     62                     return pc.load(fh, encoding=encoding, compat=True)

TypeError: invalid file: PosixPath('/home/ntaw/tmp.pickle')

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
/home/ntaw/anaconda3/lib/python3.5/site-packages/pandas/io/pickle.py in try_read(path, encoding)
     50         try:
---> 51             with open(path, 'rb') as fh:
     52                 return pkl.load(fh)

TypeError: invalid file: PosixPath('/home/ntaw/tmp.pickle')

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
/home/ntaw/anaconda3/lib/python3.5/site-packages/pandas/io/pickle.py in try_read(path, encoding)
     55             try:
---> 56                 with open(path, 'rb') as fh:
     57                     return pc.load(fh, encoding=encoding, compat=False)

TypeError: invalid file: PosixPath('/home/ntaw/tmp.pickle')

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
<ipython-input-3-d33de9a00518> in <module>()
----> 1 df2 = pd.read_pickle(path)

/home/ntaw/anaconda3/lib/python3.5/site-packages/pandas/io/pickle.py in read_pickle(path)
     66     except:
     67         if PY3:
---> 68             return try_read(path, encoding='latin1')
     69         raise
     70 

/home/ntaw/anaconda3/lib/python3.5/site-packages/pandas/io/pickle.py in try_read(path, encoding)
     59             # compat pickle
     60             except:
---> 61                 with open(path, 'rb') as fh:
     62                     return pc.load(fh, encoding=encoding, compat=True)
     63 

TypeError: invalid file: PosixPath('/home/ntaw/tmp.pickle')
```

#### Problem description

`pandas` breaks at `df = pd.read_pickle(path)`. It seems to me that this should have been solved along with issue [#11033](https://github.com/pandas-dev/pandas/issues/11033), but it evidently doesn't work for my setup.

I encountered the same issue with `pd.DataFrame.to_pickle`.

The usecase appears to be related to open issue [#14705](https://github.com/pandas-dev/pandas/issues/14705) and closed pull request [#12930](https://github.com/pandas-dev/pandas/pull/12930).

#### Expected Output

 I would expect the `pandas.read_*` functions and `pandas.DataFrame.to_*`  methods to recognise `pathlib` objects, and handle the conversion `str(path)` under the hood.

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-53-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.19.1
nose: 1.3.7
pip: 9.0.1
setuptools: 27.2.0
Cython: 0.25.1
numpy: 1.11.2
scipy: 0.18.1
statsmodels: 0.6.1
xarray: None
IPython: 5.1.0
sphinx: 1.4.8
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2016.7
blosc: None
bottleneck: 1.1.0
tables: 3.3.0
numexpr: 2.6.1
matplotlib: 1.5.3
openpyxl: 2.4.0
xlrd: 1.0.0
xlwt: 1.1.2
xlsxwriter: 0.9.3
lxml: 3.6.4
bs4: 4.5.1
html5lib: None
httplib2: 0.9.2
apiclient: 1.5.1
sqlalchemy: 1.1.3
pymysql: None
psycopg2: None
jinja2: 2.8
boto: 2.43.0
pandas_datareader: None
</details>

### LINKED ISSUE #16291: Regression in 0.20: IO functions can't handle pathlib.Path objects anymore
#### Code Sample, a copy-pastable example if possible

```python
import numpy as np
import pandas as pd
from pathlib import Path
d = pd.DataFrame(np.random.randn(25).reshape((5, 5)))
d.to_pickle('test.pickle')
d2 = pd.read_pickle(Path('test.pickle'))
```
#### Problem description

In 0.19, Pandas' IO functions accepted PathLike objects as per PEP 519. This is broken in 0.20.1.

#### Expected Output

*nothing, code should run without errors or output*

#### Actual Output

```
Traceback (most recent call last):
  File "/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/pandas/io/pickle.py", line 83, in try_read
    return read_wrapper(lambda f: pkl.load(f))
  File "/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/pandas/io/pickle.py", line 68, in read_wrapper
    return func(f)
  File "/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/pandas/io/pickle.py", line 83, in <lambda>
    return read_wrapper(lambda f: pkl.load(f))
TypeError: file must have 'read' and 'readline' attributes

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/pandas/io/pickle.py", line 88, in try_read
    lambda f: pc.load(f, encoding=encoding, compat=False))
  File "/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/pandas/io/pickle.py", line 68, in read_wrapper
    return func(f)
  File "/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/pandas/io/pickle.py", line 88, in <lambda>
    lambda f: pc.load(f, encoding=encoding, compat=False))
  File "/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/pandas/compat/pickle_compat.py", line 187, in load
    fh.seek(0)
AttributeError: 'PosixPath' object has no attribute 'seek'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/pandas/io/pickle.py", line 94, in read_pickle
    return try_read(path)
  File "/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/pandas/io/pickle.py", line 92, in try_read
    lambda f: pc.load(f, encoding=encoding, compat=True))
  File "/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/pandas/io/pickle.py", line 68, in read_wrapper
    return func(f)
  File "/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/pandas/io/pickle.py", line 92, in <lambda>
    lambda f: pc.load(f, encoding=encoding, compat=True))
  File "/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/pandas/compat/pickle_compat.py", line 187, in load
    fh.seek(0)
AttributeError: 'PosixPath' object has no attribute 'seek'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/pandas/io/pickle.py", line 83, in try_read
    return read_wrapper(lambda f: pkl.load(f))
  File "/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/pandas/io/pickle.py", line 68, in read_wrapper
    return func(f)
  File "/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/pandas/io/pickle.py", line 83, in <lambda>
    return read_wrapper(lambda f: pkl.load(f))
TypeError: file must have 'read' and 'readline' attributes

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/pandas/io/pickle.py", line 88, in try_read
    lambda f: pc.load(f, encoding=encoding, compat=False))
  File "/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/pandas/io/pickle.py", line 68, in read_wrapper
    return func(f)
  File "/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/pandas/io/pickle.py", line 88, in <lambda>
    lambda f: pc.load(f, encoding=encoding, compat=False))
  File "/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/pandas/compat/pickle_compat.py", line 187, in load
    fh.seek(0)
AttributeError: 'PosixPath' object has no attribute 'seek'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/pandas/io/pickle.py", line 97, in read_pickle
    return try_read(path, encoding='latin1')
  File "/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/pandas/io/pickle.py", line 92, in try_read
    lambda f: pc.load(f, encoding=encoding, compat=True))
  File "/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/pandas/io/pickle.py", line 68, in read_wrapper
    return func(f)
  File "/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/pandas/io/pickle.py", line 92, in <lambda>
    lambda f: pc.load(f, encoding=encoding, compat=True))
  File "/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/pandas/compat/pickle_compat.py", line 187, in load
    fh.seek(0)
AttributeError: 'PosixPath' object has no attribute 'seek'
```

#### Output of ``pd.show_versions()``

<details>
>>> pd.show_versions()
/home/mruffalo/.opt/python3.6-venv/lib/python3.6/site-packages/xarray/core/formatting.py:16: FutureWarning: The pandas.tslib module is deprecated and will be removed in a future version.
  from pandas.tslib import OutOfBoundsDatetime

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.1.final.0
python-bits: 64
OS: Linux
OS-release: 4.8.0-51-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.20.1
pytest: None
pip: 9.0.1
setuptools: 35.0.2
Cython: 0.25.2
numpy: 1.12.1
scipy: 0.19.0
xarray: 0.9.1
IPython: 6.0.0
sphinx: None
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: None
tables: 3.3.0
numexpr: 2.6.2
feather: None
matplotlib: 2.0.0
openpyxl: None
xlrd: 1.0.0
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by chris-b1**
- [x] closes #16291, closes #14932
 - [ ] tests added / passed
 - [ ] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/io/test_pickle.py [hunk 1]
```diff
@@ -299,6 +300,15 @@ def test_pickle_v0_15_2():
     tm.assert_categorical_equal(cat, pd.read_pickle(pickle_path))
 
 
+def test_pickle_path_pathlib():
+    tm._skip_if_no_pathlib()
```
Comment (jreback) [inline pandas/tests/io/test_pickle.py]: instead of making a specific test can u add a general routine like we have for round_trip_pickle

Comment (jreback) [inline pandas/tests/io/test_pickle.py]: and then use it here

ideally add tests for all of the io functions (and xfail id they don't work)

and test for localpath as well

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/io/test_pickle.py (hunk 1 above)
Comment (chris-b1) [inline pandas/tests/io/test_pickle.py]: Good point, let me know if what I just pushed is what you had in mind and I'll add to other io functions

File: pandas/tests/io/test_pickle.py [hunk 2]
```diff
@@ -299,6 +300,19 @@ def test_pickle_v0_15_2():
...
+    df = tm.makeDataFrame()
+    result = tm.round_trip_pathlib(df.to_pickle, pd.read_pickle)
+    tm.assert_frame_equal(df, result)
+
+
+def test_pickle_path_localpath():
+    tm._skip_if_no_localpath()
```
Comment (jreback) [inline pandas/tests/io/test_pickle.py]: I would move the skip routines to the actual test functions themselves (in util/testing)

File: pandas/util/testing.py [hunk 3]
```diff
@@ -117,6 +117,26 @@ def round_trip_pickle(obj, path=None):
...
+    if path is None:
+        path = '___pathlib___'
+    with ensure_clean(path) as path:
+        writer(Path(path))
+        obj = reader(Path(path))
+    return obj
+
+
+def round_trip_localpath(writer, reader, path=None):
+    from py.path import local as LocalPath
```
Comment (jreback) [inline pandas/util/testing.py]: e.g. here. 

and you need an pytimportor.skip here (for both of these).

File: pandas/tests/io/test_pickle.py (hunk 2 above)
Comment (jreback) [inline pandas/tests/io/test_pickle.py]: ignore this, you can just use a skip importer below.

File: pandas/tests/io/test_pickle.py (hunk 1 above)
Comment (jreback) [inline pandas/tests/io/test_pickle.py]: yep exactly like this!

maybe put a little docs in the testing file above these 3 tests routines.

File: pandas/io/common.py [hunk 4]
```diff
@@ -314,6 +314,8 @@ def _get_handle(path_or_buf, mode, encoding=None, compression=None,
 
     handles = list()
     f = path_or_buf
+    # Convert pathlib.Path/py.path.local or string
```
Comment (jreback) [inline pandas/io/common.py]: can you add a blank line

File: doc/source/whatsnew/v0.20.2.txt [hunk 5]
```diff
@@ -33,6 +33,8 @@ Performance Improvements
 Bug Fixes
 ~~~~~~~~~
 
+- Bug in using ``pathlib.Path`` object with io functions (:issue:`16291`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.2.txt]: localpath too?

**Review by jreback — APPROVED** [review]
fantastic. thanks! (tiny comments). merge when ready.

File: pandas/tests/io/parser/common.py [hunk 6]
```diff
@@ -679,6 +679,18 @@ def test_file(self):
...
 
+    def test_path_pathlib(self):
+        df = tm.makeDataFrame()
+        result = tm.round_trip_pathlib(df.to_csv,
+                                       lambda p: self.read_csv(p, index_col=0))
+        tm.assert_frame_equal(df, result)
+
+    def test_pickle_path_localpath(self):
+        df = tm.makeDataFrame()
+        result = tm.round_trip_pathlib(df.to_csv,
```
Comment (jorisvandenbossche) [inline pandas/tests/io/parser/common.py]: round_trip_localpath ?

**Comment by jreback** [comment]
for those ones which xfail this, can you create an issue (put all in issue). Though might be one that just needs updating.

File: pandas/tests/io/sas/test_sas7bdat.py [hunk 7]
```diff
@@ -65,6 +65,26 @@ def test_from_iterator(self):
                 tm.assert_frame_equal(df, df0.iloc[2:5, :])
                 rdr.close()
 
+    def test_path_pathlib(self):
```
Comment (jreback) [inline pandas/tests/io/sas/test_sas7bdat.py]: looks like needs xfailing here

**Comment by jreback** [comment]
can you add tests for feather-format as well (it should work I think, not sure)

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by chris-b1** [comment]
@jreback - this should be good now, looks like most of the xfails will be fixed by #16301

**Comment by TomAugspurger** [comment]
There was a failure in one job: https://travis-ci.org/pandas-dev/pandas/jobs/231516184#L2140 but that looks unrelated.

**Comment by jreback** [comment]
thanks!

**Pull request merged**